File: server/kith/api/routes/events.py
"""The one stream the interface listens to.

This replaces two endpoints. `/api/changes` said what changed and lost anything queued when a
connection dropped; `/api/activity/stream` carried the feed and re-sent its whole backlog on every
connect, duplicating lines. Two channels, two failure modes, and — since the renderer loads the
backend over plain HTTP/1.1 — two of Chromium's six connections per origin held permanently.

One endpoint, two event types, and the difference between them is not cosmetic:

* ``changed`` — a kind of thing moved. No values, ever. The client invalidates and refetches
  through the API it already uses. See `kernel/changes`.
* ``activity`` — a line of what he is doing. The value *is* the event: there is no endpoint that
  will tell you again what a tool call said thirty seconds ago, so this one carries its payload.

SSE rather than a WebSocket, deliberately. Nothing here flows the other way — the client has the
whole REST API for that — and `id:`/`Last-Event-ID` is a resumption protocol the browser and every
proxy already implement. A socket would mean writing one.

`Last-Event-ID` is the whole reason this exists in this shape. See `kernel/events` for the three
ways a cursor can be wrong and the honest answer to each; this module is that contract on the wire.
"""

from __future__ import annotations

import json
import queue

from flask import Response, request

from kith.api.blueprint import api
from kith.kernel import events

#: Seconds without an event before a keep-alive comment goes out. Short enough that a dead
#: connection surfaces as a write failure here rather than as a window that has quietly stopped
#: hearing anything, and well inside the idle timeout of anything that might sit in between.
_PING = 15


def _frame(event: events.Event) -> str:
    """One event as SSE wire text.

    The `id:` line is the contract: the client sends the last one it saw back as `Last-Event-ID`
    on reconnect, and notices a gap if two arrive out of step. `event:` is what lets one
    connection carry both kinds — `EventSource.addEventListener("activity", …)` is how SSE was
    designed to multiplex, and using it is what collapsed two sockets into one.

    The id is `<epoch>-<n>`, not a bare number, and the epoch half is what makes a restart
    detectable however late the client reconnects. SSE treats the id as opaque text and echoes it
    back verbatim, so carrying two facts in it costs nothing. See `kernel/events`.
    """
    return (
        f"id: {events.EPOCH}-{event.id}\n"
        f"event: {event.type}\n"
        f"data: {json.dumps(event.data)}\n\n"
    )


def _cursor() -> tuple[int | None, str | None]:
    """Where this client got to, and which run of the server it got there in.

    The header is the browser's own, set automatically by `EventSource` on a reconnect; the desktop
    shell sends the same header from the main process, where it holds the cursor across renderer
    reloads. A `?since=` alternative was left out on purpose: two ways to say one thing means two
    code paths to keep honest, and the header is the one the standard specifies.

    `<epoch>-<n>`. A bare number is still accepted and read as "no epoch given", which is what a
    client that predates this sends — it then gets the numeric checks alone, which is what it had.
    """
    raw = request.headers.get("Last-Event-ID", "").strip()
    if not raw:
        return None, None
    epoch, _, number = raw.rpartition("-")
    try:
        return int(number), (epoch or None)
    except ValueError:
        # A cursor we cannot read is not a cursor. Treated as a fresh connection rather than as an
        # error: the client's next fetch is what it actually needs, and a 400 on a reconnect loop
        # would be a window that never recovers.
        return None, None


@api.get("/events")
@api.doc(
    summary="What is happening",
    description="One server-sent stream: `changed` events naming what moved, `activity` lines of "
    "what he is doing. Resumable with `Last-Event-ID`.",
)
def events_stream():
    since, epoch = _cursor()
    subscription = events.subscribe(since, epoch)

    def generate():
        try:
            # An immediate byte, so the client knows the stream is *open* rather than merely
            # connecting — and so anything buffering until first byte lets go of it.
            yield ": open\n\n"

            if subscription.resync:
                # We cannot say what was missed, so we say that. The client refetches everything
                # it holds and takes the cursor we are actually at, rather than being told it is
                # up to date when it is not.
                yield _frame(events.Event(id=subscription.at, type="resync", data={}))
            else:
                for event in subscription.missed:
                    yield _frame(event)

            while True:
                try:
                    event = subscription.queue.get(timeout=_PING)
                except queue.Empty:
                    yield ": ping\n\n"
                    continue
                yield _frame(event)
        finally:
            # Reached when the client goes away, which for a generator is the next write after it
            # did. The keep-alive is what makes that happen at all on an idle stream.
            events.unsubscribe(subscription)

    response = Response(
        generate(),
        mimetype="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            # Tells nginx and anything like it not to sit on the stream waiting for a whole
            # response. Harmless when nothing is in front of us, which is the usual case.
            "X-Accel-Buffering": "no",
            #
            # `Connection: keep-alive` was here, copied from every SSE example there is, and it is
            # a real bug rather than a redundancy: `Connection` is a hop-by-hop header, which
            # PEP 3333 forbids a WSGI application from setting. Werkzeug let it through; waitress
            # raises on it, so the whole endpoint answered 500 the moment it ran on a production
            # server. It was never doing anything — the connection stays open because the response
            # is a generator that has not finished, not because of a header.
        },
    )
    # Subscribing happens above, at request time, rather than in the generator's first line — a
    # generator does not run until something consumes it, and an event published in that gap would
    # be missed silently, which is the failure this whole module exists to remove.
    #
    # Which leaves one case the `finally` cannot reach: a response closed before it was ever
    # started, whose generator therefore never runs. `call_on_close` catches that, and unsubscribe
    # is a `discard` — so the two together are safe rather than merely likely.
    response.call_on_close(lambda: events.unsubscribe(subscription))
    return response

File: desktop/src/server/events.ts
/**
 * The one event stream, held by the shell rather than by the page.
 *
 * The renderer used to open this itself — two `EventSource`s, in fact, one for what changed and
 * one for the activity feed. That works, and it costs more than it looks:
 *
 * **The connection outlived nothing.** Closing the window *hides* it, a reload replaces the
 * document, a crashed renderer is recreated — and each of those dropped the subscription and
 * started a new one with no idea where the old one got to. Every gap was silent, which is why the
 * interface kept eleven timers as a safety net. Held here, the stream and its cursor belong to the
 * *app*: the page can come and go underneath it.
 *
 * **It could not authenticate.** A browser `EventSource` cannot send headers, so the route had to
 * be exempted from the API token and gated on being same-origin instead — a hole that existed
 * purely because of a limitation in the client. Node's HTTP client has no such limitation, so the
 * shell's stream sends `X-Kith-Token` like every other call it makes.
 *
 * **It was in Chromium's socket pool.** The renderer loads the backend over plain HTTP/1.1, where
 * Chromium allows six connections per origin — and two of them were permanently spent on streams.
 * A parked turn holding a third is how a window came to sit there with every fetch queued behind a
 * full pool. This connection is Node's, and Node does not share that pool.
 *
 * `Last-Event-ID` is the whole point. See `server/kith/kernel/events.py` for the three ways a
 * cursor can be wrong; this end's job is to remember it across everything the page cannot.
 */

import * as http from "node:http";
import * as https from "node:https";

import { apiHeaders } from "./api-token";
import { reader, type ParsedEvent } from "./sse";
import { BACKEND_ORIGIN } from "../config";
import { getMainWindow } from "../window/window";

/** The channel a renderer listens on. Mirrored in `preload.ts` and nowhere else. */
export const CHANNEL = "kith:event";

/** What the page receives. The parsed event minus the raw id, which is ours to keep. */
export type ServerEvent = Omit<ParsedEvent, "raw">;

/* Reconnect backoff.
 *
 * Starts fast because the overwhelmingly common case is the server restarting under `KITH_RELOAD`
 * during development, which is over in under a second. Caps at ten, because past that the answer
 * is not a shorter interval — the server is down, and `Last-Event-ID` means waiting costs a replay
 * rather than a gap. */
const RETRY_MIN_MS = 500;
const RETRY_MAX_MS = 10_000;

/** Long enough that a keep-alive comment (every 15s server-side) always beats it. */
const IDLE_TIMEOUT_MS = 45_000;

let request: http.ClientRequest | null = null;
let retry: NodeJS.Timeout | null = null;
let wait = RETRY_MIN_MS;
let stopped = false;

/** Where we got to, exactly as the server wrote it (`<epoch>-<n>`), so it can be handed straight
 *  back as `Last-Event-ID`. Survives reconnects, renderer reloads, and the window being hidden. */
let cursor = "";

/**
 * Hand an event to our own window, and only ours.
 *
 * Not `BrowserWindow.getAllWindows()`, which is what this said first. The render service keeps
 * hidden windows open on *arbitrary web pages* — that is its whole job — and sending the app's
 * internal events into those is wrong even though they cannot read them: they have no preload, so
 * there is no `ipcRenderer` to listen with. "Cannot read it" is not a reason to send it.
 *
 * Sent while hidden, though. Closing this app's window hides it and the page keeps its React
 * state, so it has to keep hearing or it comes back showing a world from whenever it was last
 * visible.
 */
function fanOut(event: ServerEvent): void {
  const window = getMainWindow();
  if (!window || window.isDestroyed()) return;
  window.webContents.send(CHANNEL, event);
}

/** Open the stream, and keep it open. */
function connect(): void {
  if (stopped || request) return;

  const url = new URL("/api/events", BACKEND_ORIGIN);
  const client = url.protocol === "https:" ? https : http;
  const headers = apiHeaders({
    Accept: "text/event-stream",
    "Cache-Control": "no-cache",
    // Only once we have somewhere to resume from. On a first connection the page fetches its own
    // state anyway, and asking to replay from zero would hand it a backlog it has already got.
    ...(cursor ? { "Last-Event-ID": cursor } : {}),
  });

  const outgoing = client.request(
    { protocol: url.protocol, hostname: url.hostname, port: url.port, path: url.pathname, headers },
    (response) => {
      if (response.statusCode !== 200) {
        // Most likely the server is still starting, or has just been replaced by the reloader.
        response.resume();
        outgoing.destroy();
        return;
      }
      // Connected: the next drop starts its backoff from the bottom again rather than from
      // however long the last outage took to end.
      wait = RETRY_MIN_MS;
      response.setEncoding("utf8");
      const feed = reader(({ raw, ...event }) => {
        // The raw id is what goes back on the wire; the page gets everything but it.
        if (raw) cursor = raw;
        fanOut(event);
      });
      response.on("data", feed);
      response.on("end", () => outgoing.destroy());
    },
  );

  request = outgoing;
  // A stream is idle by design between events, so this guards only against a connection that has
  // died without telling us — which the server's keep-alive is what makes detectable.
  outgoing.setTimeout(IDLE_TIMEOUT_MS, () => outgoing.destroy());
  outgoing.on("error", () => {
    /* Reconnecting is the response to every error here; `close` is where that happens. */
  });
  outgoing.on("close", () => {
    request = null;
    if (stopped) return;
    if (retry) clearTimeout(retry);
    retry = setTimeout(connect, wait);
    wait = Math.min(wait * 2, RETRY_MAX_MS);
  });
  outgoing.end();
}

/**
 * Start listening, and keep the app listening for as long as it runs.
 *
 * Called once the backend answers, so the first attempt is not spent on a server that is still
 * booting — but nothing depends on that being true, because a failed attempt is just the first
 * retry.
 */
export function startEvents(): void {
  stopped = false;
  connect();
}

/** Let go, on quit. */
export function stopEvents(): void {
  stopped = true;
  if (retry) clearTimeout(retry);
  retry = null;
  request?.destroy();
  request = null;
}
